### DISCUSSION TITLE
Volume control curve

### DISCUSSION THREAD

**Discussion by vorosj** [pschatzmann/arduino-audio-tools#1614] [Q&A] [answered]
Not sure, if it's a feature or, bug. :-)
I measured the pp amplitude of a PCM5102A module. The code is simple, generator -> i2s with volume control.
Here is the result:
vol       Vpp (mV)
0.1       118
0.2       238
0.3       348
0.4       470
0.5       578
0.6     1640
0.7     2660
0.8     3670
0.9     4780
1.0     5780
The increment is higher from 0.5. So the control is nor linear, nor logaritmic, but two linear sections with a break point.
Of course I can live with it, and easily correct it with a table, however I suspect it's not the intended behavior.
Or I missed something?

`#include "Arduino.h"
/**
 * @file streams-generator-i2s.ino
 * @author Phil Schatzmann
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-generator-i2s/README.md 
 * @copyright GPLv3
 */
 
#include "AudioTools.h"

AudioInfo info(44100, 2, 16);
SineWaveGenerator<int16_t> sineWave(32000);                // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<int16_t> sound(sineWave);             // Stream generated from sine wave
I2SStream out; 
VolumeStream volume(out);
//StreamCopy copier(out, sound);                             // copies sound into i2s
StreamCopy copier(volume, sound);

long prevMillis;
int freq = 500;
float vol = 1;

// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  while(!Serial);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // start I2S
  Serial.println("starting I2S...");
  auto config = out.defaultConfig(TX_MODE);
  
  config.pin_bck = 2;
  config.pin_ws = 15;
  config.pin_data = 4;
    
  config.copyFrom(info); 
  out.begin(config);

  // Setup sine wave
  sineWave.begin(info, N_B4);
  sineWave.setFrequency(freq);

  volume.begin(config); // we need to provide the bits_per_sample and channels
  volume.setVolume(vol);

  delay(2000);
  Serial.println("started...");
  prevMillis = millis();
}

// Arduino loop - copy sound to out 
void loop() {
  copier.copy();

  if (prevMillis < (millis()-2000)){
    prevMillis = millis();
    //freq+=10;
    //sineWave.setFrequency(freq);
  vol+=0.1;
  if (vol >1){
     vol = 1;
  }
    Serial.println(vol);
    //volume.setVolume(vol);
  }
}`

**Comment by pschatzmann** [comment, marked as answer]
Did you read the corresponding Wiki ?
This matches pretty much with what has been described and it is up to you to select whatever you want...

**Comment by fufu930** [comment]
Kudos for sharing a detailed PCM5102A Vpp metrics!! 🙃🙂
